src/typescript-service.ts

```diff
@@ -83,13 +84,161 @@ export default class TypeScriptService {
         return res;
     }
 
+    private ensuredFilesForHoverAndDefinition = new Map<string, Promise<void>>();
+
+    private ensureFilesForHoverAndDefinition(uri: string): Promise<void> {
+        if (this.ensuredFilesForHoverAndDefinition.get(uri)) {
+            return this.ensuredFilesForHoverAndDefinition.get(uri);
+        }
+
+        const fileName: string = util.uri2path(uri);
+        const absFileName = "/" + fileName;
+
+        const promise = this.projectManager.ensureModuleStructure().then(() => {
+            return this.projectManager.ensureFiles([absFileName]);
+        }).then(() => {
+            return this.projectManager.refreshConfigurations();
+        }).then(() => {
+            return this.projectManager.getConfiguration(fileName).get().then((config) => {
+                const contents = config.moduleResolutionHost().readFile(fileName);
+                const info = ts.preProcessFile(contents, true, true);
+                const compilerOpt = config.host.getCompilationSettings();
+
+                const importDirs = new Set<string>();
+                for (const imp of info.importedFiles) {
```

Comment:
We should test the following case:
A imports B, B imports C. A.a is an instance of B.foo which is {bar: string; baz: C.qux}
What we'll see as A.a's hover if C is not in the `info.importedFiles`?